### PR TITLE
feat(outcomes): new orgstats endpoints, outcomes snuba module

### PR DESCRIPTION
**Pull request by JoshFerge**
- API description, architecture: https://www.notion.so/sentry/Sentry-outcomes-consumer-module-backend-OrgStats-changes-ee78192ee4f443be9be66009968de06f
- Adds an endpoint and snuba module to do queries against our outcomes data in clickhouse.
- The architecture/querying is very similar to the sessions_v2 endpoint, and shares code and extends where applicable.
- It differs primarily in how you filter: instead of using the `query` field, filters for `outcome`, `category`, and `reason` are passed directly as query params. This seemed more sensible than trying to extend our querying grammar to work with outcomes
    - it looks like sessions has some auto-aggregation and dataset switching on the snuba side, which outcomes does not have, so had to add a small amount of logic to handle this query side.
- I typed the `outcomes.py` file and added it to the `mypy.ini` file, would love thoughts on how I typed it. also added minimal typing elsewhere to have type compatible imports.

Things to test in staging:

- [x] performance of queries in large orgs, project wise and event wise. 
- [x] performance of queries for minute rollups
- [x] performance of queries with different filters applied
- [ ] test issue where you can't see all projects on stats page

### TODO
- [ ] mark old outcomes TSDB module as deprecated through code comment or otherwise
- [x] final small TODOs in code

### CODE REVIEW

File: src/sentry/api/urls.py [hunk 1]
```diff
@@ -1210,6 +1214,21 @@
                     OrganizationStatsEndpoint.as_view(),
                     name="sentry-api-0-organization-stats",
                 ),
+                url(
```
Comment (JoshFerge) [inline src/sentry/api/urls.py:1217]: need to decide on how we do versioning with this API. we intend to keep the old one around in case some customers are using it to generate reports. for now, `_v2` is added there.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            groupby=["category", "time", "outcome", "reason"],
+            aggregations=[["sum", "times_seen", "times_seen"], ["sum", "quantity", "quantity"]],
+            filter_keys={"org_id": [organization.id]},
+            orderby=["time"],
+        )
+
+        response = {
+            "statsErrors": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsTransactions": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsAttachments": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:53]: Nit: instead of the constant `DEFAULT_TS_VAL`, defining

```python
def create_default_ts_value():
    return {...}
```

would let you just write `defaultdict(create_default_ts_value)`, avoiding the use of both `lambda` and `deepcopy`.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            "times_seen": row["times_seen"],
+        }
+    return obj
+
+
+def nested_update(d, u):
+    # https://stackoverflow.com/a/3233356
+    for k, v in u.items():
+        if isinstance(v, collections.abc.Mapping):
+            d[k] = nested_update(d.get(k, {}), v)
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:197]: Recursing to arbitrary depth seems squirrelly to me. In the places where this is called, would it be possible to unpack the appropriate number of levels instead?

It looks like this is meant to explore a structure like
```python
{
    "accepted": {"quantity": 0, "times_seen": 0},
    "filtered": {"quantity": 0, "times_seen": 0},
    "dropped": {
        "overQuota": {"quantity": 0, "times_seen": 0},
        "spikeProtection": {"quantity": 0, "times_seen": 0},
        "other": {"quantity": 0, "times_seen": 0},
    },
}
```
and hit a `{"quantity": 0, "times_seen": 0}` object whether it's one or two layers deep. See my other comment about introducing classes regarding that.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+        response = {
+            "statsErrors": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsTransactions": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsAttachments": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+        }
+        for row in result:
+            nested_update(
+                response[CATEGORY_NAME_MAP[row["category"]]][row["time"]], map_row_to_format(row)
+            )
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:148]: I'm seeing a few readability issues here, which I think could be improved in part by introducing one or more classes to model the data that you're building to put into the response, and only converting it to a raw serializable value at the end.

If I'm getting this correctly, there are three layers of structure to model. The topmost is the data categories, the second is the structure of `DEFAULT_TS_VAL`, and the third is `{"quantity": 0, "times_seen": 0}`. If I were writing this from scratch and trying to maximize clarity, I'd probably write one class for each of those three.

One advantage is that the second class could just be a flat structure of five values ("accepted", "filtered", "overQuota", "spikeProtection", "other"), where you only worry about wrapping the last three into the `dropped` dict at the final serialization step.

If you're interested in undertaking this as a refactoring, I'll pair-program on it if you like.

Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:148]: yes, good call on all of this. will give modeling this w/ classes a shot and making it more readable, will reach out for a pair-program if i'm feeling anything is ambiguous. thank you!

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        response = StatsResponse(start, end, rollup)
+        for row in result:
+            stat_to_update = response.get(row["category"])
+            stat_to_update.update(row)
+
+        return Response(response.build_fields())
+
+
+# TODO: good names for these classes
+class StatsResponse:
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_stats_v2.py:110]: Other parts of the application have this kind of logic in `sentry.api.serializers`. There are also some useful base classes in `sentry.api.serializers.snuba` that help with zerofilling timeseries results.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        self.other = StatMeasure(0, 0)
+        self.time = time
+
+    def serialize(self):
+        return {
+            "accepted": {
+                "quantity": self.accepted.quantity,
+                "times_seen": self.accepted.times_seen,
+            },
+            "filtered": {"quantity": self.filtered.quantity, "times_seen": self.filtered.quantity},
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_stats_v2.py:174]: Could you use the datacategory to decide whether to serialize quantity or times_seen as the output `value`? That would make frontend code simpler and reduce your payload size as we don't need both for each category.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        )
+
+
+class OrganizationProjectStatsDetails(ProjectEndpoint, OrganizationEndpoint):
+    def get(self, request, project):
+        start, end, rollup = get_date_range_rollup_from_params(request.GET, "1h", round_range=True)
+
+        # TODO: see if there's a better way to get the user's projects
+        project_list = []
+        team_list = Team.objects.get_for_user(organization=project.organization, user=request.user)
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_stats_v2.py:83]: You can/should use `OrganizationEndpoint.get_projects()` for this. It handles both kinds of 'all' and organization permissions around open team membership which would be needed here too.

File: src/sentry/snuba/outcomes.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        groupby=groupby,
+        aggregations=aggregations,
+        rollup=rollup,
+        filter_keys=filter_keys,
+        orderby=orderby,
+        dataset=Dataset.Outcomes,
+    )
+
+    for row in result["data"]:
+        row["category"] = (
```
Comment (markstory) [inline src/sentry/snuba/outcomes.py:24]: This is likely simpler than trying to cook up a case statement that will yield a similar result in clickhouse.

File: src/sentry/snuba/outcomes.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+from sentry.utils.snuba import (
+    raw_query,
+)
+from sentry.snuba.dataset import Dataset
+from sentry_relay import DataCategory
+
+
+def query(groupby, start, end, rollup, aggregations, orderby, fields=None, filter_keys=None):
+
+    # TODO: if roll up >= use hourly, else use raw
```
Comment (markstory) [inline src/sentry/snuba/outcomes.py:10]: You might also want to move the rollup constraint here instead of in the endpoint. That will help with not making invalid queries if this module is used elsewhere in the future.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+class OrganizationStatsEndpointV2(OrganizationEndpoint):
+    def get(self, request, organization):
+        start, end, rollup = get_date_range_rollup_from_params(request.GET, "1h", round_range=True)
+        result = outcomes.query(
+            start=start,
+            end=end,
+            rollup=rollup,
+            groupby=["category", "time", "outcome", "reason"],
+            aggregations=[["sum", "times_seen", "times_seen"], ["sum", "quantity", "quantity"]],
+            filter_keys={"org_id": [organization.id]},
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_stats_v2.py:33]: Should this also filter by the projects you're allowed to see, but not expose that information in the breakdowns?

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+        response = StatsResponse(start, end, rollup)
+        for row in result:
+            stat_to_update = response.get(row["category"])
+            stat_to_update.update(row)
+
+        # response.zerofill(start, end, rollup)
+        return Response(response.build_fields())
+
+
+class OrganizationProjectStatsIndex(OrganizationEndpoint):
```
Comment (markstory) [inline src/sentry/api/endpoints/organization_stats_v2.py:46]: You might be able to combine this with the Project specific endpoint if the client sends a projectid parameter in the request. The base organization endpoint's `get_projects()` will interpret no project parameter as 'all projects I have membership in'. You can also send -1 from the client to access all projects you have access to. Another benefit of using that approach is that project permissions are enforced more consistently with the rest of the product.

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 5 above)
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:110]: would it be okay for these classes to live in `serializers/models` or do we limit those to just ones with corresponding django serializers? i see in a few other places we create a top level folder for endpoints w/ helper classes included in the folder

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 6 above)
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:174]: definitely.

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 7 above)
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:83]: 👍🏼

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    def __init__(self, start, end, rollup):
+        self.errors = TimeSeriesValues(start, end, rollup)
+        self.transactions = TimeSeriesValues(start, end, rollup)
+        self.attachments = TimeSeriesValues(start, end, rollup)
+
+    _GETTERS = {
+        DataCategory.ERROR: (lambda s: s.errors),
+        DataCategory.TRANSACTION: (lambda s: s.transactions),
+        DataCategory.ATTACHMENT: (lambda s: s.attachments),
+    }
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:120]: I take it this is based on the pattern I used in the `UsageStat` class in `billing_usage.py`? It's really only worth it if you want to expose the regular attributes while also offering the `get` method (in my case, I started with the attributes and was forced to bolt on the `get` afterward).

If you can get by with only the `get` (which seems to be the case here at a glance), I'd recommend skipping the attributes entirely and just having an internal dict, keyed by the categories. That simplifies a few things:

```python
    def __init__(self, start, end, rollup):
        self._values = {k: TimeSeriesValues(start, end, rollup) for k in CATEGORY_NAME_MAP}

    def get(self, category):
        return self._values[category]

    def __iter__(self):
        return iter(self._values.items())
```
etc.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+            else:
+                self.other.quantity = row["quantity"]
+                self.other.times_seen = row["quantity"]
+
+        elif row["outcome"] == Outcome.ACCEPTED:
+            self.accepted.quantity = row["quantity"]
+            self.accepted.times_seen = row["times_seen"]
+        elif row["outcome"] == Outcome.FILTERED:
+            self.filtered.quantity = row["quantity"]
+            self.filtered.times_seen = row["times_seen"]
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:206]: Suggested refactoring. A few more lines but it trims down the repetition.

```python
    def update(self, row):
        def find_measure():
            if row["outcome"] == Outcome.RATE_LIMITED:
                if row["reason"] in {"usage_exceeded", "grace_period"}:
                    return self.over_quota, True
                elif row["reason"] == "smart_rate_limit":
                    return self.spike_protection, False
                else:
                    return self.other, False
            elif row["outcome"] == Outcome.ACCEPTED:
                return self.accepted, True
            elif row["outcome"] == Outcome.FILTERED:
                return self.filtered, True
            else:
                return None, None  # or raise an error?

        measure_to_update, use_times_seen = find_measure()
        if measure_to_update:
            measure_to_update.quantity = row["quantity"]
            measure_to_update.times_seen = row["times_seen" if use_times_seen else "quantity"]
```

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            data_by_time[obj["time"]].append(obj)
+        else:
+            data_by_time[obj["time"]] = [obj]
+    for key in range(start, end, rollup):
+        if key in data_by_time and len(data_by_time[key]) > 0:
+            rv = rv + data_by_time[key]
+            data_by_time[key] = []
+        else:
+            val = MeasureValue(key)
+            rv[key] = val
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:237]: It seems like `zerofill` is doing a lot of work up front and it's a little hard to follow. Is it only filling in empty entries in case the corresponding timestamps don't later show up in the data? If so, is there a good way to push that to the serialization step, starting instead with empty lists/dicts?

Just a thought. I'll take your word for it if you say that's more trouble than it's worth.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+            },
+            "time": self.time,
+        }
+
+    def update(self, row):
+        if row["outcome"] == Outcome.RATE_LIMITED:
+            if row["reason"] in {"usage_exceeded", "grace_period"}:
+                self.over_quota.quantity = row["quantity"]
+                self.over_quota.times_seen = row["times_seen"]
+            elif row["reason"] in "smart_rate_limit":
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:194]: Should be `row["reason"] == "smart_rate_limit"`?

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+        self.times_seen = times_seen
+
+
+class MeasureValue:
+    def __init__(self, time):
+        self.accepted = StatMeasure(0, 0)
+        self.filtered = StatMeasure(0, 0)
+        self.over_quota = StatMeasure(0, 0)
+        self.spike_protection = StatMeasure(0, 0)
+        self.other = StatMeasure(0, 0)
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:165]: Nit: I'd write it as `StatMeasure()`.

**Review by RyanSkonnord — COMMENTED** [review]
👍 on the OO refactor. I hope you agree that it's a bit easier to work with this way.

Names are hard. 😉

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 14 above)
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:237]: yeah, how it is now isn't ideal. i think ideally we move it into the outcomes snuba query, and this code can gracefully create an object and this complexity can be removed from the class.

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 15 above)
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:194]: yup 👍🏼

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 16 above)
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:165]: 👍🏼

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 13 above)
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:206]: 👍🏼 will add that in. going to set quantity and times_seen to the row values here and return a singular value in serialization based on the DataCategory, so helper function will just return the one value.

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 5 above)
Comment (markstory) [reply src/sentry/api/endpoints/organization_stats_v2.py]: I'm fine with them being in serializers, as that directory also contains DRF serializers for request parsing. The top level directory for endpoints was something we tried where parts of the product could be separate django applications, but it didn't really get that much traction.

File: src/sentry/snuba/outcomes.py [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        raise NotImplementedError()
+
+
+class QuantityField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
+        return ["quantity"]
+
+    def extract_from_row(
+        self, row: Dict[str, Union[Any, int]], group: Optional[Dict[str, Any]] = None
```
Comment (JoshFerge) [inline src/sentry/snuba/outcomes.py:68]: have to do this weird Union to get mypi to like having this function returning an int. wonder what the best way to to this is.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        except NoProjects:
+            raise NoProjects("No projects available")
+
+        return QueryDefinition(
+            request.GET,
+            params,
+        )
+
+    @contextmanager
+    def handle_query_errors(self):
```
Comment (JoshFerge) [inline src/sentry/api/endpoints/organization_stats_v2.py:61]: this bit of TODO was copied over from the organization_sessions endpoint. its probably worth it to refactor and introduce a new endpoint base that sessions and this one can use for these inherited methods, but didn't want this PR to balloon, so keeping it as is for now.

File: src/sentry/snuba/outcomes.py [hunk 19]
```diff
@@ -0,0 +1,330 @@
+from typing import Any, Dict, List, Optional, Tuple
+from abc import ABC, abstractmethod
+from django.http import QueryDict
+from sentry.utils.snuba import raw_query
+from .dataset import Dataset
+from sentry_relay import DataCategory
+
+from sentry.snuba.sessions_v2 import (
```
Comment (JoshFerge) [inline src/sentry/snuba/outcomes.py:8]: there is a potential to move some of this (and other funcs within other snuba modules) into a common that the endpoints can import / inherit from, but abstracting out didn't feel worth it here, so just importing from sessions directly.

File: src/sentry/snuba/outcomes.py [hunk 20]
```diff
@@ -0,0 +1,330 @@
...
+class Dimension(SimpleGroupBy, ABC):
+    @abstractmethod
+    def resolve_filter(self, raw_filter: List[str]) -> List[DataCategory]:
+        """
+        Based on the input filter, map it back to the clickhouse representation
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def map_row(self, row: Dict[str, Any]) -> None:
```
Comment (evanh) [inline src/sentry/snuba/outcomes.py:99]: Here, since `row` is modified by the function `Dict` or `MutableMapping` works.

File: src/sentry/snuba/outcomes.py [hunk 21]
```diff
@@ -0,0 +1,330 @@
...
+    **DIMENSION_MAP,
+    "project": SimpleGroupBy("project_id", "project"),
+}
+
+TS_COL = "time"
+
+
+def get_filter(
+    query: QueryDict, params: Dict[Any, Any]
+) -> Tuple[List[Tuple[str, str, List[Any]]], Dict[str, Any]]:
```
Comment (evanh) [inline src/sentry/snuba/outcomes.py:173]: This is where a NamedTuple for a single Condition would clean this up a bit. Or if returning conditions and filter keys in a tuple happens a lot, wrapping that in a NamedTuple.

File: src/sentry/snuba/outcomes.py [hunk 22]
```diff
@@ -0,0 +1,330 @@
...
+
+        query_groupby = set()
+        for groupby in self.groupby:
+            query_groupby.update(groupby.get_snuba_groupby())
+        self.query_groupby = list(query_groupby)
+
+        self.conditions, self.filter_keys = get_filter(query, params)
+
+
+def run_outcomes_query(query: QueryDefinition) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
```
Comment (evanh) [inline src/sentry/snuba/outcomes.py:248]: Adding `ResultSet = List[Dict[str, Any]]` would clean up a lot of these functions.

**Review by evanh — COMMENTED** [review]
Some typing suggestions. Way to be a pioneer!

File: src/sentry/snuba/outcomes.py [hunk 23]
```diff
@@ -0,0 +1,330 @@
...
+- `project`
+- `outcome`
+- `reason`
+- `category`
+"""
+
+
+class Field(ABC):
+    @abstractmethod
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
```
[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py]: got it! makes sense, thanks.

File: src/sentry/snuba/outcomes.py [hunk 24]
```diff
@@ -0,0 +1,330 @@
...
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        raise NotImplementedError()
+
+
+class QuantityField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
+        return ["quantity"]
+
+    def extract_from_row(self, row: Dict[str, Any], group: Optional[Dict[str, Any]] = None) -> int:
+        if row is None:
```
[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py:71]: aha this one was not quite right. added an optional wrapper for that. thanks for the catch!

File: src/sentry/snuba/outcomes.py (hunk 20 above)
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py]: 👍🏼

File: src/sentry/snuba/outcomes.py (hunk 21 above)
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py]: Tried to make it a namedtuple, but looks like our snuba json serializer likes to convert namedtuples to dicts, which made snuba unhappy about that. Figured it's best not to jump down that rabbit hole so just using a regular tuple type variable, definitely looks cleaner. thanks!

File: src/sentry/snuba/outcomes.py (hunk 22 above)
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py]: 100%. updated! looks much better now.

File: src/sentry/snuba/outcomes.py (hunk 17 above)
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py]: fixed, went ahead and just typecasting the return value to `int`, which is essentially a noop here. h/t @RyanSkonnord

File: src/sentry/api/urls.py (hunk 1 above)
Comment (dbennett-sentry) [reply src/sentry/api/urls.py:1222]: Hi! Long time listener, first time caller. Can we check the logs to confirm anyone outside Sentry is directly calling the original API?

Comment (JoshFerge) [reply src/sentry/api/urls.py:1222]: looks like 11 orgs have called this API in the last 90 days in sentry SaaS using the best proxy i could for non-sentry-webpage calls.

File: src/sentry/snuba/outcomes.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (evanh) [reply src/sentry/snuba/outcomes.py:71]: If you run mypy with the `--warn-unreachable` flag I think it should have surfaced this as well.

File: src/sentry/snuba/outcomes.py [hunk 25]
```diff
@@ -0,0 +1,339 @@
...
+
+    result_totals = _format_rows(result["data"], query)
+    result_timeseries = _format_rows(result_timeseries["data"], query)
+    return result_totals, result_timeseries
+
+
+def _format_rows(rows: ResultSet, query: QueryDefinition) -> ResultSet:
+    category_grouping: Dict[str, Any] = {}
+
+    def _group_row(row: Dict[str, Any]) -> None:
```
Comment (JoshFerge) [inline src/sentry/snuba/outcomes.py:292]: will definitely want to see performance of this in staging. there are some easier optimizations to be had with this function (we only need to do it on rows category = error, security, default). would be an issue in orgs with large project cardinality. additionally, it could be done in the massaging step so we're not looping through twice.

Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py:292]: there is also the option of doing the groupby in the snuba query instead of here, which is also an option.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+        try:
+            # TODO: this context manager should be decoupled from `OrganizationEventsEndpointBase`?
+            with super().handle_query_errors():
```
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/organization_stats_v2.py:64]: It looks like `OrganizationEventsEndpointBase.handle_query_errors` is logically a static method (that is, its body does not refer to `self`). 👍 both to pulling it into a new base class and to deferring it outside this PR. But, if you're looking for another way, it might be fine as a top-level function if you want to avoid inheritance entirely.

File: src/sentry/snuba/outcomes.py [hunk 27]
```diff
@@ -0,0 +1,341 @@
...
+        if row is None:
+            return 0
+        return int(row["times_seen"])
+
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        if dataset == Dataset.Outcomes:
+            return ["sum", "times_seen", "times_seen"]
+        else:
+            # RawOutcomes doesnt have times_seen, do a count instead
+            return ["count()", "", "times_seen"]
```
Comment (RyanSkonnord) [inline src/sentry/snuba/outcomes.py:95]: What's the deal with the three aggregation values? Do the three positions have any significance? Is it documented anywhere?

A dataclass with three fields could improve on `List[str]`, but probably not worth the extra lines without 3.8's dataclass feature.

If the value is constrained to always have the same length, I think a tuple matches Python conventions better than a list. You can do `Tuple[str, str, str]` as the type hint, right?

File: src/sentry/snuba/outcomes.py [hunk 28]
```diff
@@ -0,0 +1,341 @@
...
+class CategoryDimension(Dimension):
+    def resolve_filter(self, raw_filter: Sequence[str]) -> List[DataCategory]:
+        resolved_categories = set()
+        for category in raw_filter:
+            # combine DEFAULT, ERROR, and SECURITY as errors.
+            # see relay: py/sentry_relay/consts.py and relay-cabi/include/relay.h
+            if DataCategory.parse(category) == DataCategory.ERROR:
+                resolved_categories.update(DataCategory.error_categories())
+            else:
+                resolved_categories.add(DataCategory.parse(category))
```
Comment (RyanSkonnord) [inline src/sentry/snuba/outcomes.py:123]: I think I've seen this pattern around elsewhere. Any chance a `DataCategory` class method like the following could be used from more than one place?
```python
def parse_categories(cls, category):
    dc = cls.parse(category)
    return cls.error_categories() if dc == cls.ERROR else [dc]
```

File: src/sentry/snuba/outcomes.py (hunk 27 above)
Comment (evanh) [reply src/sentry/snuba/outcomes.py:95]: Just FYI, dataclasses are available as a third party library in Python <3.8.

> You can do Tuple[str, str, str] as the type hint, right?

Just a note that since the return value isn't a Tuple mypy wouldn't like that type hinting.

File: src/sentry/api/endpoints/organization_stats_v2.py [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+            with sentry_sdk.start_span(op="outcomes.endpoint", description="run_outcomes_query"):
+                result_totals, result_timeseries = run_outcomes_query(query)
+            with sentry_sdk.start_span(op="sessions.endpoint", description="run_sessions_query"):
+                result = massage_outcomes_result(query, result_totals, result_timeseries)
+            return Response(result, status=200)
+
+    def build_outcomes_query(self, request, organization):
+        for param in request.GET:
+            if param not in VALID_PARAMETERS:
+                raise InvalidParams(f'Invalid parameter: "{param}"')
```
[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:49]: 👍🏼 will go ahead and remove this bit. and the various parameters end up being validated by the helper functions.

File: src/sentry/snuba/outcomes.py [hunk 30]
```diff
@@ -0,0 +1,341 @@
...
+    InvalidField,
+)
+from sentry.utils.outcomes import Outcome
+from datetime import datetime
+from sentry.search.utils import InvalidQuery
+
+"""
+The new Outcomes API defines a "metrics"-like interface which is can be used in
+a similar way to "sessions" and "discover"
+# "Metrics"
```
[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py:21]: 👍🏼

File: src/sentry/snuba/outcomes.py [hunk 31]
```diff
@@ -0,0 +1,341 @@
...
+        start=query.start,
+        end=query.end,
+        groupby=query.query_groupby,
+        aggregations=query.aggregations,
+        rollup=query.rollup,
+        filter_keys=query.filter_keys,
+        conditions=query.conditions,
+        selected_columns=query.query_columns,
+        referrer="outcomes.totals",
+        # orderby=query.orderby, TODO: add orderby?
```
[the remark this answers is not in the document]
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py:271]: we end up sorting in the post processing function, will keep an eye on this when testing performance

File: src/sentry/api/endpoints/organization_stats_v2.py (hunk 26 above)
Comment (JoshFerge) [reply src/sentry/api/endpoints/organization_stats_v2.py:64]: will add this to my few pieces of tech debt bill from this PR, but making a static method seems 👌🏼

File: src/sentry/snuba/outcomes.py (hunk 27 above)
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py:95]: @RyanSkonnord that piece is documented in `SnubaQueryParams` in `utils/snuba.py` : 
> `aggregations` a list of (aggregation_function, column, alias) tuples to be passed to the query.

Definitely agree this could be more obvious with types / NamedTuples / DataClasses. Tried to use some NamedTuples for some of the types but we JSON serialize NamedTuples as JSON objects instead of an array, so didn't work out there. For now, i've converted this return type to Tuple[str,str,str] as it is more descriptive.

File: src/sentry/snuba/outcomes.py (hunk 28 above)
Comment (JoshFerge) [reply src/sentry/snuba/outcomes.py]: agreed it could be useful elsewhere, but yes also agree out of scope of this changeset.

**Review by RyanSkonnord — APPROVED** [review]
Implementation looks good to me. Someone else might still take a pass to check feature/integration correctness.
